### ISSUE TITLE
Error code should be 500 when a public key cannot be extracted from the mp.jwt.verify.publickey.location property

### ISSUE THREAD

**Issue by SoniaZaldana** [smallrye/smallrye-jwt#372]
Hi all, 

I am currently working on resolving the WildFly issue [WFLY-13164](https://issues.redhat.com/browse/WFLY-13164). I noticed that `JWTHttpAuthenticationMechanism` sends the 401 status code when validating a request throws an exception.

However, when this exception is caused by failing to extract a public key from the `mp.jwt.verify.publickey.location` property (see `DefaultJwtTokenParser` and `KeyLocationResolver`), I was wondering whether it would be more appropriate to send a 500 status code instead.

Perhaps we can throw a 500 status code when we catch a `ParseException`? 

Thanks for your help, 
Sonia

**Comment by sberyozkin** [comment]
Hi @SoniaZaldana Looks like the issue is related to the corrupted key inlined with `mp.jwt.public.key` property - let me have a look

**Comment by sberyozkin** [comment]
I think what happened is that since we started supporting EllipticCurve keys the early conversion to `RSAPublicKey` became delayed which can explain the regression; I'll try to add a test and a fix today

**Comment by sberyozkin** [comment]
Hi @SoniaZaldana and @darranl 

I've been looking at this issue for a while today - and have come to a conclusion that maybe a problem lies at the container level which translates `ParserException` ? `smallrye-jwt` itself sits above the HTTP transport, it only reports the exceptions :-).

For example, if we look at [DefaultJwtTokenParser](https://github.com/smallrye/smallrye-jwt/blob/2.3.1/implementation/src/main/java/io/smallrye/jwt/auth/principal/DefaultJWTTokenParser.java) starting from the 2.3.1 tag, you'll see that if the key is unresolvable, then `ParseException` wrapping `UnresolvableKeyException` cause [is reported](https://github.com/smallrye/smallrye-jwt/blob/2.3.1/implementation/src/main/java/io/smallrye/jwt/auth/principal/DefaultJWTTokenParser.java#L186). Nothing has changed there - you can trace the history - some updates happened in that block but nothing fundamentally changed.

@SoniaZaldana Can you please check what might've changed at the WildFly level processing `ParseException` ?

**Comment by sberyozkin** [comment]
Maybe the problem is still at the smallrye-jwt level, let me look further

**Comment by sberyozkin** [comment]
No, `KeyLocationResolver` already has a unit test verifying `UnresolvableKeyException` is reported if the location is wrong and I've just tried locally a unit test with the wrong content (which would come from `mp.jwt.public.key`):

```
@Test
    public void testLoadeyWithWrongContent() throws Exception {

        expectedEx.expect(UnresolvableKeyException.class);
        when(signature.getHeaders()).thenReturn(headers);
        when(headers.getStringHeaderValue(JsonWebKey.KEY_ID_PARAMETER)).thenReturn("1");
        JWTAuthContextInfo context = new JWTAuthContextInfo();
        context.setIssuedBy("http://abc");
        context.setPublicKeyContent("12345");
        KeyLocationResolver keyLocationResolver = new KeyLocationResolver(context);
        keyLocationResolver.resolveKey(signature, emptyList());
    }
```

This one also passes for me.
So @SoniaZaldana, can you please check what happens on the container end when exception is reported ?

**Comment by SoniaZaldana** [comment]
Hi @sberyozkin thanks for looking into it :-)

For context, WildFly is using the `2.0.13` tag for smallrye-jwt. This is due to some refactoring that happened in the next tag. Therefore, I am limited to testing my container with this tag. 

As you mentioned, essentially nothing has changed in the `DefaultJwtTokenParser`. It wraps an `UnresolvableKeyException` in a `ParseException`. I can confirm I see this behaviour when I run my container. 

The issue lies in [JWTHttpAuthenticationMechanism#validateRequest](https://github.com/smallrye/smallrye-jwt/blob/master/implementation/jwt-auth/src/main/java/io/smallrye/jwt/auth/mechanism/JWTHttpAuthenticationMechanism.java#L65). When we attempt to parse the bearer token, `DefaultJwtTokenParser` sends the `UnresolvableKeyException` wrapped in the `ParseException`, but the exception handling in `JWTHttpAuthenticationMechanism#validateRequest` sends the status code [401](https://github.com/smallrye/smallrye-jwt/blob/master/implementation/jwt-auth/src/main/java/io/smallrye/jwt/auth/mechanism/JWTHttpAuthenticationMechanism.java#L82). 

However, given this is a server error, I was wondering whether it would be more appropriate for the exception handling to send the 500 status code when we are dealing with a wrapped `UnresolvableKeyException`. 

I hope this clarifies things a bit, and I apologize if my initial description was a little misleading.

**Comment by sberyozkin** [comment]
Hi @SoniaZaldana Oh great, nice investigation, thanks; yes, definitely, lets do it. Are you OK with opening your first `smallrye-jwt` PR ? That would be great. I suppose at this stage it would be fine if you could confirm the fix against WildFly - perhaps in the future we can also have this mechanism tested locally - I think Darran has started on this path with this pending PR: #256

**Comment by sberyozkin** [comment]
Oh I can open a PR against the current code and you'd cherry-pick or just copy it as we've split and moved the modules a bit

**Comment by sberyozkin** [comment]
@SoniaZaldana FYI, in smallrye-jwt 3.0 I'll focus on initializing the local key early - this is what was done in 2.0.13 I think, and I'll do for both inlined keys and the keys in the file resources.
But the `JWTHttpAuthenticationMechanism` fix you have proposed is a good one - it will ensure 500 is returned consistently. The consistency problem is related to the case where we have for example a local or HTTPS based JWK set - with more than one key in a JWK format - in such cases, unless a smallrye-jwt specific property specifying a key id is set, the key will always be calculated dynamically - using the incoming token `kid` property to select the key.
So the mechanism fix will ensure 500 is always returned correctly in case of the unavailable or corrupt key.

I'll need to release `smallrye-jwt-2.4.2` early next week, Wed or so, I'll need to fix a few more issues for 2.4.2 so your help with this PR will be appreciated. See how it goes - like I said I can help with a PR myself.
Cheers, Sergey

**Comment by SoniaZaldana** [comment]
Hi @sberyozkin, I can work on submitting the PR. Thanks for all the help!
